inappmessaging/src/main/java/com/rakuten/tech/mobile/inappmessaging/runtime/utils/SharePreferencesUtil.kt

```diff
@@ -1,17 +1,79 @@
 package com.rakuten.tech.mobile.inappmessaging.runtime.utils
 
 import android.content.Context
+import android.content.SharedPreferences
 import androidx.security.crypto.EncryptedSharedPreferences
 import androidx.security.crypto.MasterKey
+import timber.log.Timber
+import java.io.File
+import java.security.KeyStore
 
 internal object SharePreferencesUtil {
-    fun generateKey(context: Context) = MasterKey.Builder(context).setKeyScheme(MasterKey.KeyScheme.AES256_GCM).build()
-
-    fun createSharedPreference(context: Context, key: MasterKey, account: String) = EncryptedSharedPreferences.create(
-                context,
-                account,
-                key,
-                EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
-                EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
-        )
+
+    @SuppressWarnings("TooGenericExceptionCaught", "ReturnCount", "LongMethod", "NestedBlockDepth")
+    fun createSharedPreference(
+        context: Context,
+        account: String,
+        key: MasterKey? = null,
+        shouldRetry: Boolean = true,
+        timber: Timber.Tree = Timber.tag(TAG)
+    ): SharedPreferences {
+        if (isEncryptedEnabled(context)) {
+            try {
+                val sharedPref = EncryptedSharedPreferences.create(
+                        context,
+                        account,
+                        key ?: MasterKey.Builder(context).setKeyScheme(MasterKey.KeyScheme.AES256_GCM).build(),
+                        EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
+                        EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
+                )
+                setEncryptedEnabled(context, true)
+                return sharedPref
+            } catch (e: Exception) {
+                timber.e(e)
+                if (shouldRetry) {
+                    try {
+                        removeMasterKey(context)
+                        return createSharedPreference(context, account, shouldRetry = false)
+                    } catch (ex: Exception) {
+                        timber.e(ex)
+                    }
+                }
+            }
+        }
+        return getNormalSharedPreference(context, account)
+    }
+
+    private fun removeMasterKey(context: Context) {
+        KeyStore.getInstance("AndroidKeyStore").run {
+            load(null)
+            deleteEntry(MasterKey.DEFAULT_MASTER_KEY_ALIAS)
+        }
+
+        context.filesDir.parent?.let { parentPath ->
+            File(parentPath + File.separator + "shared_prefs" +
+                    File.separator + MasterKey.DEFAULT_MASTER_KEY_ALIAS + ".xml").delete()
+        }
+    }
+
+    private fun getNormalSharedPreference(context: Context, account: String): SharedPreferences {
+        setEncryptedEnabled(context, false)
+        return context.getSharedPreferences(PREFS + "_$account", Context.MODE_PRIVATE)
```

Comment:
Is normal SharedPref safe enough to store userId?

Reply:
No other choice here if it failed in creating the encrypted shared pref

Reply:
I hope that's ok. We could use hashed userId like on iOS but that might be a future improvement